### PR TITLE
Variables State

### PR DESCRIPTION
**Pull request by ephread**
This PR completes #193.

Thanks for reviewing!

### CODE REVIEW

File: src/VariablesState.ts [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+		this.variableChangedEventCallbacks.forEach((callback) => {
+			callback(variableName, newValue);
+		});
+	}
+
+	get batchObservingVariableChanges(){
+		return this._batchObservingVariableChanges;
+	}
+	set batchObservingVariableChanges(value: boolean){
+		value = !!value; // TODO: Might have become useless with TypeScript.
```
Comment (y-lohse) [inline src/VariablesState.ts]: Yup, useless now.

File: src/VariablesState.ts [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+					}
+				}
+			}
+		}
+	}
+
+	get callStack(){
+		return this._callStack;
+	}
+	set callStack(callStack: any /* CallStack */){
```
Comment (y-lohse) [inline src/VariablesState.ts]: Why can't we use the `CallStack` type here ?

File: src/VariablesState.ts [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+			this.SetGlobal(variableName, val);
+		}
+	}
+
+	constructor(callStack: any /* CallStack */, listDefsOrigin: ListDefinitionsOrigin){
+		this._globalVariables = new Map();
+		this._callStack = callStack;
+		this._listDefsOrigin = listDefsOrigin;
+
+		// TODO should we discourage the use of proxies since they're non-portable
```
Comment (y-lohse) [inline src/VariablesState.ts]: Can you elaborate more on this please?

File: src/VariablesState.ts [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+			// It first tries to access the objects own property, and if none is
+			// found it delegates the call to the $ method, defined below
+			let p = new Proxy(this, {
+				get(target: any, name){
+					return (name in target) ? target[name] : target.$(name);
+				},
+				set(target: any, name, value){
+					if (name in target) target[name] = value;
+					else target.$(name, value);
+					return true; // returning a falsy value make the trap fail
```
Comment (y-lohse) [inline src/VariablesState.ts]: I wonder what I meant by that, but it doesn't look very sound.

File: src/VariablesState.ts (hunk 3 above)
Comment (ephread) [reply src/VariablesState.ts]: I don't know much about proxies, this is most likely a comment inserted by @NQNStudios, so you'd have to ask them!

File: src/VariablesState.ts (hunk 2 above)
Comment (ephread) [reply src/VariablesState.ts]: Definitely, I guess that Callstack wasn't merged yet when the PR was first worked on.

File: src/VariablesState.ts (hunk 4 above)
Comment (ephread) [reply src/VariablesState.ts]: Maybe we should rework the entire proxy thing later?

Comment (y-lohse) [reply src/VariablesState.ts]: Yah, we'll see :)

**Comment by y-lohse** [comment]
Awesome, thanks for the last changes!

**Pull request merged**